content/post/20220520-quantum-compile-part2.md

```diff
@@ -0,0 +1,228 @@
+
+Coalton compiles to Common Lisp. A statically typed language being compiled to a dynamically typed language seems like a raw deal. We went through the effort of ascertaining the type of every term in our program, and we'll just throw it away?
+
+Of course not. Fortunately, Common Lisp allows the programmer to annotate types. The subset of Common Lisp's type system that permits optimizations for speedy code is, for the most part, limited to monomorphic machine, record, and array types. Polymorphism is not expressed in Common Lisp's type system in a way that is practically useful.
+
+Since Coalton knows the type of everything, it can annotate these types in the resulting Common Lisp code. When the code is monomorphic, the Common Lisp compiler will be able to produce fast code.
+
+The story is a little more nuanced with polymorphic code. Coalton supports functions that have either *parametric polymorphism* (one function implementation that supports many types), *ad hoc polymorphism* (a single function name that has many implementations depending on the argument and result types), or both.
+
+A function like `flip` is parametrically polymorphic:
+
+```lisp
+> (type-of 'flip)
+∀ :A :B :C. ((:A → :B → :C) → :B → :A → :C)
+```
+
+Flip can take _any_ two-argument function (taking types `:A` and `:B`), and produce a new function with the arguments flipped (thus taking `:B` and `:A`). It's not too much of a problem compiling this kind of parametric polymorphism in efficient Common Lisp.
+
+Similarly, a function like `+` is ad-hoc-polymorphic:
+
+```lisp
+> (type-of '+)
+∀ :A. NUM :A ⇒ (:A → :A → :A)
+```
+
+This is because how `+` works depends on the type of `:A`. The implementation of `Integer` addition is certainly different from the implementation of `AlgNum` addition. Since we need to be able to work with different implementations of `+` corresponding to different data types at run-time, optimization is more difficult.
+
+Coalton implements ad hoc polymorphism with something that roughly looks like a "vtable"-approach: it passes around tables—known to functional compiler engineers as *dictionaries*—of concrete method implementations to the callees, which they can draw from and call. So, for example, a function that uses the `+` operator will, at run-time, be suppled a table saying "this is how you implement `+` for an `integer`" when `+` is happens to actually get called on an `integer`. Since this is a dynamic thing, we pay the relatively steep penalties of constructing these dictionaries and making indirect calls into them.
+
+As such, we've implemented optimizations for a few common sources of inefficiency. We'll describe two of them: _monomorphization_ and _specialization_.
+
+### Monomorphization: Opportunistically eliminating polymorphism when it's not needed
```

Comment:
This section would be clearer if it discussed direct calling methods. That would explain how dictionaries are actually eliminated, and why `quad-poly` gets a performance improvement w/ just a type signature.

Also maybe show the call site differences between the original call to square in `quad-mono` and the call to square after the monomorphizer.

Reply:
I'm going to delete this stuff and put it in a separate PR at another time